File: mcp/src/minecraft/femtocraft/api/IPowerContainer.java
/*******************************************************************************
 * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 * Itszuvalex@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 ******************************************************************************/

package femtocraft.api;

import femtocraft.managers.research.EnumTechLevel;

/**
 * @author Itszuvalex
 */
public interface IPowerContainer {

    /**
     * @param level EnumTechLevel of power of container
     * @return True if can accept power of that level
     */
    public boolean canAcceptPowerOfLevel(EnumTechLevel level);

    /**
     * @return EnumTechLevel of container
     */
    public EnumTechLevel getTechLevel();

    /**
     * @return Current storage amount of container
     */
    public int getCurrentPower();

    /**
     * @return Max storage amount of container - used for percentage
     * approximations during charging
     */
    public int getMaxPower();

    /**
     * @return Actual fill percentage - for things like damage values, etc.
     */
    public float getFillPercentage();

    /**
     * @return Fill percentage for purposes of charging - allows tanks and
     * whatnot to trick pipes into filling them I.E. return
     * getFillPercentage() < .25f ? getFillPercentage() : .25f;
     */
    public float getFillPercentageForCharging();

    /**
     * @return Fill percentage for purposes of output - allows tanks and other
     * TileEntities to trick pipes into not pulling all of their power.
     */
    public float getFillPercentageForOutput();

    /**
     * @return True if container has room and can accept charging from direction
     * @from false otherwise
     */
    public boolean canCharge();

    /**
     * @param amount Amount attempting to charge.
     * @return Total amount of @amount used to fill internal tank.
     */
    public int charge(int amount);

    /**
     * @param amount Amount of power to drain from internal storage
     * @return True if all power was consumed, false otherwise. This anticipates
     * all or nothing behavior.
     */
    public boolean consume(int amount);

}

File: src/main/scala/com/itszuvalex/femtocraft/api/power/IPowerContainer.scala
package com.itszuvalex.femtocraft.api.power

import com.itszuvalex.femtocraft.api.EnumTechLevel

/**
 * @author Itszuvalex
 *
 *         Most generic power containment interface.  This simply provides the interface into the power system without worrying about sidedness.
 *
 *         Expected Behaviors ->
 *         Implemented by an Item =>  A PowerStackWrapper made with any itemstack with this item should be valid.
 *         Anywhere else =>   Any general case on your own.
 */
trait IPowerContainer {
  /**
   * @param level EnumTechLevel of power of container
   * @return True if can accept power of that level
   */
  def canAcceptPowerOfLevel(level: EnumTechLevel): Boolean

  /**
   * @return EnumTechLevel of container
   */
  def getTechLevel: EnumTechLevel

  /**
   * @return Current storage amount of container
   */
  def getCurrentPower: Int

  /**
   * @return Max storage amount of container - used for percentage approximations during charging
   */
  def getMaxPower: Int

  /**
   * @return Actual fill percentage - for things like damage values, etc.
   */
  def getFillPercentage: Float

  /**
   * @return Fill percentage for purposes of charging - allows tanks and whatnot to trick pipes into filling them I.E.
   *         return getFillPercentage() < .25f ? getFillPercentage() : .25f;
   */
  def getFillPercentageForCharging: Float

  /**
   * @return Fill percentage for purposes of output - allows tanks and other TileEntities to trick pipes into not
   *         pulling all of their power.
   */
  def getFillPercentageForOutput: Float

  /**
   * @return True if container has room and can accept charging false otherwise
   */
  def canCharge: Boolean

  /**
   * @param amount Amount attempting to charge.
   * @return Total amount of @amount used to fill internal tank.
   */
  def charge(amount: Int): Int

  /**
   * @param amount Amount of power to drain from internal storage
   * @return True if all power was consumed, false otherwise. This anticipates all or nothing behavior.
   */
  def consume(amount: Int): Boolean
}
